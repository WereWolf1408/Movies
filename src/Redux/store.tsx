import { configureStore } from '@reduxjs/toolkit';
import { useDispatch } from 'react-redux';
import movies from './reducers/moviesReducer';
import modalWindows from './reducers/modalWindowsReducer';
import additionalStore from './reducers/additionalReducer';

export const store = configureStore({
  reducer: {
    movies,
    additionalStore,
    modalWindows,
  },
});

export type RootState = ReturnType<typeof store.getState>;
export const AppDispatch = () => useDispatch<typeof store.dispatch>();
