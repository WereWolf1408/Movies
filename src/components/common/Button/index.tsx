import React from 'react';
const classnames = require('classnames');

import './style.less';

const CLASSES = {
  NETFLIX_BUTTON: 'netflix-app__button',
};

interface ButtonProps {
  (props: { text: string; classes?: string; clickHandler?: () => void }): JSX.Element;
}

const Button: ButtonProps = ({ text, classes, clickHandler = null }) => {
  return (
    <div
      onClick={clickHandler}
      className={classnames(CLASSES.NETFLIX_BUTTON, classes)}
    >
      {text}
    </div>
  );
};

export default Button;