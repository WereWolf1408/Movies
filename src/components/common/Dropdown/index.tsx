import React, { useState } from 'react';
import { DropdownCallbackProps } from '@utils/interfaces';

import './style.less';
import classNames from 'classnames';

const CLASSES = {
  NETFLIX_APP_DROPDOWN: 'netflix-app__dropdown',
  NETFLIX_APP_DROPDOWN_OPTIONS: 'netflix-app__dropdown-options',
  NETFLIX_APP_DROPDOWN_FIELD: 'netflix-app__dropdown-field',
  NETFLIX_APP_DROPDOWN_CHECKBOX: 'netflix-app__dropdown-checkbox',
  NETFLIX_APP_DROPDOWN_CHECKBOX_LABEL: 'netflix-app__dropdown-checkbox-label',
};

interface DropdownProps {
  (props: {
    options: Array<string>;
    callback?: DropdownCallbackProps;
    dropdownType: 'simple' | 'multiline';
  }): JSX.Element;
}

const Dropdown: DropdownProps = ({ callback, options, dropdownType }) => {
  const [selected, setSelected] = useState<string>(options[0]);
  const [isOpen, setIsOpen] = useState(false);
  const [optionsHolder, setOptionsHolder] = useState([options[0]]);

  const clickHandler = (option: string) => {
    setSelected(option);
    setIsOpen(false);
    callback && callback(option);
  };

  const multilineClickHandler = (
    option: string,
    event: React.ChangeEvent<HTMLInputElement>
  ) => {
    const { checked } = event.currentTarget;
    const modifyArr = (): Array<string> => {
      if (checked) {
        optionsHolder.push(option);
        return [...optionsHolder];
      }
      const filteredArr = optionsHolder.filter((value) => value !== option);
      return filteredArr;
    };
    const changedArr = modifyArr();
    setOptionsHolder(changedArr);
    setSelected(changedArr.toString());
    callback && callback(changedArr);
  };

  const selectedOptionClickHandler = () => {
    setIsOpen((prevVal) => !prevVal);
  };

  const renderDropdownOutputElement = () => {
    if (dropdownType === 'simple') {
      return <span onClick={selectedOptionClickHandler}>{selected}</span>;
    }
    return (
      <input
        className={'netflix-app__input'}
        readOnly={true}
        onClick={selectedOptionClickHandler}
        value={selected}
      />
    );
  };

  return (
    <div
      className={classNames(CLASSES.NETFLIX_APP_DROPDOWN, {
        [CLASSES.NETFLIX_APP_DROPDOWN_FIELD]: dropdownType === 'multiline',
      })}
    >
      {renderDropdownOutputElement()}
      {isOpen && (
        <div className={CLASSES.NETFLIX_APP_DROPDOWN_OPTIONS}>
          {options.map((item, index) => {
            return dropdownType === 'simple' ? (
              <span
                key={index}
                className={classNames({ ['active']: item === selected })}
                onClick={() => clickHandler(item)}
              >
                {item}
              </span>
            ) : (
              <div
                key={index}
                className={classNames({ ['active']: optionsHolder.includes(item) })}
              >
                <label
                  htmlFor=""
                  className={CLASSES.NETFLIX_APP_DROPDOWN_CHECKBOX_LABEL}
                >
                  <input
                    className={CLASSES.NETFLIX_APP_DROPDOWN_CHECKBOX}
                    type="checkbox"
                    value={item}
                    checked={optionsHolder.includes(item)}
                    onChange={(event) => multilineClickHandler(item, event)}
                  />
                  {item}
                </label>
              </div>
            );
          })}
        </div>
      )}
    </div>
  );
};

export default Dropdown;
