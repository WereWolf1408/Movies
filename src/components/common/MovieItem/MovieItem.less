@import '@styles/colors.less';

.netflix-app__movie {
  display: flex;
  flex-direction: column;
  max-width: 31.9rem;
  max-height: 50.7rem;
  margin: 0 5.5rem 3.7rem 0;
  color: @color-white;

  &-image {
    height: 100%;
    width: 100%;
    object-fit: cover;
  }

  &-year-name {
    display: flex;
    justify-content: space-between;
    margin-top: 1.3rem;
    font-weight: 500;
    font-size: 1.8rem;
    line-height: 2.2rem;
    opacity: .7;

    :nth-child(2) {
      border: 1px solid @color-dusty-gray;
      padding: .4rem .8rem;
      border-radius: .4rem;
    }
  }

  &-genre {
    margin-top: 0.8rem;
    opacity: .5;
  }
}
