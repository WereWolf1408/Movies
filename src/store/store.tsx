import { configureStore } from '@reduxjs/toolkit';
import { useDispatch } from 'react-redux';
import { rootReducer } from './rootReducer';
import { additionalReducer } from './additionalReducer';

export const store = configureStore({
  reducer: {
    mainStore: rootReducer,
    additionalStore: additionalReducer,
  },
});

export type RootState = ReturnType<typeof store.getState>;
export const AppDispatch = () => useDispatch<typeof store.dispatch>();
