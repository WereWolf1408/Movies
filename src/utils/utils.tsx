export const printMessage = (value: string) => {
  console.log(value);
};
