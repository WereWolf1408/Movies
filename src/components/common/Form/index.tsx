import React, { Dispatch, SetStateAction, useState } from 'react';
import {
  DateInput,
  InputWithLabel,
  DropdownInput,
  TextAreaWithLabel,
} from '../SmartInputs';
import Button from '../Button';
import { formDataReceived } from '@utils/interfaces';

import './style.less';

const CLASSES = {
  NETFLIX_APP_MODAL_WINDOW_FORM: 'netflix-app__modal-window-form',
  NETFLIX_APP_MODAL_WINDOW_FORM_INPUT_SECTION:
    'netflix-app__modal-window-form-input-section',
  NETFLIX_APP_MODAL_WINDOW_FORM_INPUT_BLOCK:
    'netflix-app__modal-window-form-input-block',
  NETFLIX_APP__INPUT_BLOCK: 'netflix-app__input-block',
  NETFLIX_APP_MODAL_WINDOW_FORM_BUTTONS_SECTION:
    'netflix-app__modal-window-form-buttons-section',
};

type FormEvent = React.FormEvent<HTMLInputElement>;

type SetFormFieldProps = (
  event: React.FormEvent<HTMLInputElement>,
  callback: Dispatch<SetStateAction<string>>
) => void;

interface FormProps {
  (props: {
    resetButtonClickHandler: () => void;
    submitClickHandler: (opt: formDataReceived) => void;
  }): JSX.Element;
}

//Consider usage of useForm hook in conjunction with yup() or any other joi.
//JOI - in simple vords - validation scheme

const Form: FormProps = ({ resetButtonClickHandler, submitClickHandler }) => {
  const [title, setTitle] = useState('');
  const [movieURL, setMovieURL] = useState('');
  const [genre, setGenre] = useState(['']);
  const [releaseDate, setReleaseDate] = useState('');
  const [rating, setRating] = useState<number>(0);
  const [runTime, setRunTime] = useState('');

  const setFormField: SetFormFieldProps = (event, callback) => {
    const { value } = event.currentTarget;
    callback(value);
  };

  const titleChangeHandler = (event: FormEvent) => {
    setFormField(event, setTitle);
  };

  const movieURLChangeHandler = (event: FormEvent) => {
    setFormField(event, setMovieURL);
  };

  const genreChangeHandler = (value: Array<string>) => {
    setGenre(value);
  };

  const releaseDateChangeHandler = (event: FormEvent) => {
    setFormField(event, setReleaseDate);
  };

  const ratingChangeHandler = (event: FormEvent) => {
    const { value } = event.currentTarget;
    setRating(parseInt(value));
  };

  const runTimeChangeHandler = (event: FormEvent) => {
    setFormField(event, setRunTime);
  };

  const overviewChangeHandler = (event: React.FormEvent<HTMLTextAreaElement>) => {};

  const subminButtonClickHandler = () => {
    submitClickHandler({
      title,
      movieURL,
      genre,
      releaseDate,
      rating,
      runTime,
    });
  };

  return (
    <form className={CLASSES.NETFLIX_APP_MODAL_WINDOW_FORM}>
      <div className={CLASSES.NETFLIX_APP_MODAL_WINDOW_FORM_INPUT_SECTION}>
        <InputWithLabel
          changeHandler={titleChangeHandler}
          placeholder={'1123'}
          labelText={'Title'}
          value={title}
          classes={CLASSES.NETFLIX_APP__INPUT_BLOCK}
        />
        <InputWithLabel
          changeHandler={movieURLChangeHandler}
          placeholder={'movie url'}
          labelText={'movie url'}
          value={movieURL}
          classes={CLASSES.NETFLIX_APP__INPUT_BLOCK}
        />
        <DropdownInput
          labelText={'genre'}
          options={['comedy', 'horor', 'crime']}
          changeHandler={genreChangeHandler}
          classes={CLASSES.NETFLIX_APP__INPUT_BLOCK}
        />
        <DateInput
          labelText={'release date'}
          changeHandler={releaseDateChangeHandler}
          value={releaseDate}
          classes={CLASSES.NETFLIX_APP__INPUT_BLOCK}
        />
        <InputWithLabel
          changeHandler={ratingChangeHandler}
          placeholder={'1123'}
          labelText={'rating'}
          value={rating.toString()}
          classes={CLASSES.NETFLIX_APP__INPUT_BLOCK}
        />
        <InputWithLabel
          changeHandler={runTimeChangeHandler}
          placeholder={'runtime'}
          labelText={'runtime'}
          value={runTime}
          classes={CLASSES.NETFLIX_APP__INPUT_BLOCK}
        />
        <TextAreaWithLabel
          changeHandler={overviewChangeHandler}
          placeholder={'Movie description'}
          labelText={'runtime'}
          value={runTime}
          classes={CLASSES.NETFLIX_APP__INPUT_BLOCK}
        />
      </div>
      <div className={CLASSES.NETFLIX_APP_MODAL_WINDOW_FORM_BUTTONS_SECTION}>
        <Button clickHandler={resetButtonClickHandler} text={'reset'} />
        <Button clickHandler={subminButtonClickHandler} text={'submit'} />
      </div>
    </form>
  );
};

export default Form;