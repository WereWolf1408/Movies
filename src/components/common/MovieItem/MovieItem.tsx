import React, { useContext, useEffect, useState } from 'react';
import { SingleMovieProps } from '@utils/interfaces';
import { ToolsIcon } from '@utils/utils';

import './MovieItem.less';
import { CloseIcon } from '../../../utils/utils';
import { NetflixAppContext } from '../../../Context/Context';

const CLASSES = {
  NETFLIX_APP_MOVIE: 'netflix-app__movie',
  NETFLIX_APP_MOVIE_iMAGE: 'netflix-app__movie-image',
  NETFLIX_APP_MOVIE_YEAR_NAME: 'netflix-app__movie-year-name',
  NETFLIX_APP_MOVIE_GENRE: 'netflix-app__movie-genre',
  NETFLIX_APP_MOVIE_TOOLS: 'netflix-app__movie-tools',
  NETFLIX_APP_MOVIE_TOOLS_WINDOW: 'netflix-app__movie-tools-window',
  NETFLIX_APP_MOVIE_TOOLS_CLOSE: 'netflix-app__movie-tools-close',
  NETFLIX_APP_MOVIE_TOOLS_ITEM: 'netflix-app__movie-tools-item',
};

interface MovieItemProps {
  (props: { item: SingleMovieProps }): JSX.Element;
}

export const MovieItem: MovieItemProps = ({ item: { src, title, year, genre } }) => {
  //is it ok to use Context here!?
  const { movieDetail, showDetails, showEditMoviePopup } = useContext(NetflixAppContext);
  const [showToolsOption, setShowToolsOption] = useState(false);

  //temporary solution, later these functions will be moved in global context
  const clickHandler = () => {
    setShowToolsOption((prevProp) => !prevProp);
  };

  const closeIconClickHandler = () => {
    setShowToolsOption(false);
  };

  const imageClickHandler = () => {
    movieDetail.handler(title, src, genre, year);
    showDetails.handler(true);
  };

  const editOptionClickHandler = () => {
    movieDetail.handler(title, src, genre, year);
    showEditMoviePopup.handler(true);
  };

  useEffect(() => {
    console.log('movie detal props');
    console.log(movieDetail);
  });

  return (
    <div className={CLASSES.NETFLIX_APP_MOVIE}>
      <ToolsIcon
        classes={CLASSES.NETFLIX_APP_MOVIE_TOOLS}
        clickHandler={clickHandler}
      />
      {showToolsOption && (
        <div className={CLASSES.NETFLIX_APP_MOVIE_TOOLS_WINDOW}>
          <CloseIcon
            classes={CLASSES.NETFLIX_APP_MOVIE_TOOLS_CLOSE}
            clickHandler={closeIconClickHandler}
          />
          <div
            className={CLASSES.NETFLIX_APP_MOVIE_TOOLS_ITEM}
            onClick={editOptionClickHandler}
          >
            Edit
          </div>
          <div className={CLASSES.NETFLIX_APP_MOVIE_TOOLS_ITEM}>Delete</div>
        </div>
      )}
      <img
        className={CLASSES.NETFLIX_APP_MOVIE_iMAGE}
        src={src}
        alt=""
        onClick={imageClickHandler}
      />
      <div className={CLASSES.NETFLIX_APP_MOVIE_YEAR_NAME}>
        <span>{title}</span>
        <span>{year}</span>
      </div>
      <span className={CLASSES.NETFLIX_APP_MOVIE_GENRE}>{genre}</span>
    </div>
  );
};
