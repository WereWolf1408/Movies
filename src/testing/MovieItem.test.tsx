import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import { store } from '../store/store';
import { MovieItem } from '../components/common/MovieItem';
import { mockMoview } from '../utils/utils';

describe('MovieItem Component', () => {
  it('---> shapshot test:', () => {
    const mockOpenFunction = jest.fn();
    const mockCloseFunction = jest.fn();
    const { asFragment } = render(
      <Provider store={store}>
        <MovieItem
          item={mockMoview}
          showMovieOptions={false}
          closeMovieOptionsHandler={mockCloseFunction}
          openMovieOptionsHandler={mockOpenFunction}
        />
      </Provider>
    );
    expect(asFragment).toMatchSnapshot();
  });

  it(`should call callback functions:`, () => {
    const mockOpenFunction = jest.fn(() => {
      console.log(`open function callback was called`);
    });
    const mockCloseFunction = jest.fn();
    const { getByText, getByRole } = render(
      <Provider store={store}>
        <MovieItem
          item={mockMoview}
          showMovieOptions={false}
          closeMovieOptionsHandler={mockCloseFunction}
          openMovieOptionsHandler={mockOpenFunction}
        />
      </Provider>
    );
    userEvent.click(getByRole('tools'));
    expect(mockOpenFunction.mock.calls.length).toBe(1);
  });
});
