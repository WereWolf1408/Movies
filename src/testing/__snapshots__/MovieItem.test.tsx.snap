// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieItem Component ---> shapshot test: 1`] = `[Function]`;
